# Loading Movies Data

In [131]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import os, time, json, math
import tmdbsimple as tmdb
from tqdm.notebook import tqdm_notebook

from sqlalchemy import create_engine
from sqlalchemy.types import *
import pymysql

In [132]:
pymysql.install_as_MySQLdb()

username = ''
password = ''

connection = f'mysql+pymysql://{username}:{password}@localhost/movies'
engine = create_engine(connection)

In [114]:
#title_basic = ("https://datasets.imdbws.com/title.basics.tsv.gz")
#title_akas = ("https://datasets.imdbws.com/title.akas.tsv.gz")
#title_ratings = ("https://datasets.imdbws.com/title.ratings.tsv.gz")

In [2]:
basics = pd.read_csv("Data/title_basics.csv.gz", low_memory=False)

In [3]:
basics.head()

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2  tt0069049     movie                         The Other Side of the Wind   
3  tt0088751     movie                                  The Naked Monster   
4  tt0094859     movie                                         Chief Zabu   

                               originalTitle  isAdult  startYear  endYear  \
0                             Kate & Leopold        0       2001      NaN   
1  El Tango del Viudo y Su Espejo Deformante        0       2020      NaN   
2                 The Other Side of the Wind        0       2018      NaN   
3                          The Naked Monster        0       2005      NaN   
4                                 Chief Zabu        0       2016      NaN   

   runtimeMinutes                  genres  
0             118  Comedy,Fantasy,Romance  
1              70                   Drama  
2             122                   Drama  
3             100    Comedy,Horror,Sci-Fi  
4              74                  Comedy

In [4]:
basics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94229 entries, 0 to 94228
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tconst          94229 non-null  object 
 1   titleType       94229 non-null  object 
 2   primaryTitle    94229 non-null  object 
 3   originalTitle   94229 non-null  object 
 4   isAdult         94229 non-null  int64  
 5   startYear       94229 non-null  int64  
 6   endYear         0 non-null      float64
 7   runtimeMinutes  94229 non-null  int64  
 8   genres          94229 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 6.5+ MB


In [5]:
akas = pd.read_csv("Data/title_akas.csv.gz", low_memory=False)

In [6]:
akas.head()

titleId  ordering                   title region language        types  \
0  tt0000001         6              Carmencita     US      NaN  imdbDisplay   
1  tt0000002         7  The Clown and His Dogs     US      NaN          NaN   
2  tt0000005        10        Blacksmith Scene     US      NaN  imdbDisplay   
3  tt0000005         1     Blacksmithing Scene     US      NaN  alternative   
4  tt0000005         6     Blacksmith Scene #1     US      NaN  alternative   

              attributes  isOriginalTitle  
0                    NaN              0.0  
1  literal English title              0.0  
2                    NaN              0.0  
3                    NaN              0.0  
4                    NaN              0.0

In [7]:
rating = pd.read_csv("Data/title_ratings.csv.gz", low_memory=False)

In [8]:
rating.head()

tconst  averageRating  numVotes
0  tt0000001            5.7      1905
1  tt0000002            5.9       256
2  tt0000005            6.2      2517
3  tt0000006            5.2       173
4  tt0000007            5.4       783

# Cleaning Data

In [121]:
basics = basics.replace({'\\N':np.nan})

In [75]:
basics.head()

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2  tt0069049     movie                         The Other Side of the Wind   
3  tt0079644     movie                                      November 1828   
4  tt0088751     movie                                  The Naked Monster   

                               originalTitle  isAdult  startYear  endYear  \
0                             Kate & Leopold        0       2001      NaN   
1  El Tango del Viudo y Su Espejo Deformante        0       2020      NaN   
2                 The Other Side of the Wind        0       2018      NaN   
3                              November 1828        0       2001      NaN   
4                          The Naked Monster        0       2005      NaN   

   runtimeMinutes                  genres  
0             118  Comedy,Fantasy,Romance  
1              70                   Drama  
2             122                   Drama  
3             140               Drama,War  
4             100    Comedy,Horror,Sci-Fi

In [123]:
akas = akas.replace({'\\N':np.nan})

In [125]:
rating = rating.replace({'\\N':np.nan})

In [127]:
basics.dropna(subset=['runtimeMinutes'], inplace=True)

In [128]:
basics.dropna(subset=['genres'], inplace=True)

In [134]:
basics = basics.loc[basics['titleType'] == 'movie']

In [151]:
basics = basics.loc[basics['genres'] != 'Documentary']

In [165]:
basics = basics.loc[basics['startYear'].between('2000','2022', inclusive=True)]

C:\Users\jonny\AppData\Local\Temp\ipykernel_18588\1918233108.py:1: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  basics = basics.loc[basics['startYear'].between('2000','2022', inclusive=True)]


In [169]:
akas = akas.loc[akas['region'] == 'US']

In [85]:
keepers = basics['tconst'].isin(akas['titleId'])
keepers

0          True
1          True
2          True
3         False
4          True
          ...  
163953     True
163954     True
163955    False
163956     True
163957    False
Name: tconst, Length: 163958, dtype: bool

In [86]:
basics = basics[keepers]
basics

tconst titleType  \
0       tt0035423     movie   
1       tt0062336     movie   
2       tt0069049     movie   
4       tt0088751     movie   
7       tt0094859     movie   
...           ...       ...   
163950  tt9914942     movie   
163951  tt9915872     movie   
163953  tt9916170     movie   
163954  tt9916190     movie   
163956  tt9916362     movie   

                                             primaryTitle  \
0                                          Kate & Leopold   
1       The Tango of the Widower and Its Distorting Mi...   
2                              The Other Side of the Wind   
4                                       The Naked Monster   
7                                              Chief Zabu   
...                                                   ...   
163950                             Life Without Sara Amat   
163951                               The Last White Witch   
163953                                      The Rehearsal   
163954                                          Safeguard   
163956                                              Coven   

                                    originalTitle  isAdult  startYear  \
0                                  Kate & Leopold        0       2001   
1       El Tango del Viudo y Su Espejo Deformante        0       2020   
2                      The Other Side of the Wind        0       2018   
4                               The Naked Monster        0       2005   
7                                      Chief Zabu        0       2016   
...                                           ...      ...        ...   
163950                 La vida sense la Sara Amat        0       2019   
163951                  My Girlfriend is a Wizard        0       2019   
163953                                   O Ensaio        0       2019   
163954                                  Safeguard        0       2020   
163956                                   Akelarre        0       2020   

        endYear  runtimeMinutes                     genres  
0           NaN             118     Comedy,Fantasy,Romance  
1           NaN              70                      Drama  
2           NaN             122                      Drama  
4           NaN             100       Comedy,Horror,Sci-Fi  
7           NaN              74                     Comedy  
...         ...             ...                        ...  
163950      NaN              74                      Drama  
163951      NaN              97       Comedy,Drama,Fantasy  
163953      NaN              51                      Drama  
163954      NaN              95  Action,Adventure,Thriller  
163956      NaN              92              Drama,History  

[94229 rows x 9 columns]

In [87]:
keepers2 = rating['tconst'].isin(akas['titleId'])
keepers2

0          True
1          True
4          True
5          True
6          True
           ... 
1262939    True
1262946    True
1262947    True
1262951    True
1262965    True
Name: tconst, Length: 475727, dtype: bool

In [88]:
rating = rating[keepers2]
rating

tconst  averageRating  numVotes
0        tt0000001            5.7      1905
1        tt0000002            5.9       256
4        tt0000005            6.2      2517
5        tt0000006            5.2       173
6        tt0000007            5.4       783
...            ...            ...       ...
1262939  tt9916204            8.1       242
1262946  tt9916348            8.5        17
1262947  tt9916362            6.4      4814
1262951  tt9916428            3.8        14
1262965  tt9916720            5.1       245

[475727 rows x 3 columns]

In [106]:
#checking if 'startYear' is numeric
basics[basics['startYear'] == 2020]

tconst titleType  \
0       tt0035423     movie   
1       tt0062336     movie   
2       tt0069049     movie   
4       tt0088751     movie   
7       tt0094859     movie   
...           ...       ...   
163950  tt9914942     movie   
163951  tt9915872     movie   
163953  tt9916170     movie   
163954  tt9916190     movie   
163956  tt9916362     movie   

                                             primaryTitle  \
0                                          Kate & Leopold   
1       The Tango of the Widower and Its Distorting Mi...   
2                              The Other Side of the Wind   
4                                       The Naked Monster   
7                                              Chief Zabu   
...                                                   ...   
163950                             Life Without Sara Amat   
163951                               The Last White Witch   
163953                                      The Rehearsal   
163954                                          Safeguard   
163956                                              Coven   

                                    originalTitle  isAdult  startYear  \
0                                  Kate & Leopold        0       2020   
1       El Tango del Viudo y Su Espejo Deformante        0       2020   
2                      The Other Side of the Wind        0       2020   
4                               The Naked Monster        0       2020   
7                                      Chief Zabu        0       2020   
...                                           ...      ...        ...   
163950                 La vida sense la Sara Amat        0       2020   
163951                  My Girlfriend is a Wizard        0       2020   
163953                                   O Ensaio        0       2020   
163954                                  Safeguard        0       2020   
163956                                   Akelarre        0       2020   

        endYear  runtimeMinutes                     genres  
0           NaN             118     Comedy,Fantasy,Romance  
1           NaN              70                      Drama  
2           NaN             122                      Drama  
4           NaN             100       Comedy,Horror,Sci-Fi  
7           NaN              74                     Comedy  
...         ...             ...                        ...  
163950      NaN              74                      Drama  
163951      NaN              97       Comedy,Drama,Fantasy  
163953      NaN              51                      Drama  
163954      NaN              95  Action,Adventure,Thriller  
163956      NaN              92              Drama,History  

[94229 rows x 9 columns]

# Exporting Data

In [94]:
basics.to_csv("Data/title_basics.csv.gz",compression='gzip',index=False)

In [93]:
akas.to_csv("Data/title_akas.csv.gz",compression='gzip',index=False)

In [92]:
rating.to_csv("Data/title_ratings.csv.gz",compression='gzip',index=False)

In [100]:
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94229 entries, 0 to 163956
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tconst          94229 non-null  object 
 1   titleType       94229 non-null  object 
 2   primaryTitle    94229 non-null  object 
 3   originalTitle   94229 non-null  object 
 4   isAdult         94229 non-null  int64  
 5   startYear       94229 non-null  int64  
 6   endYear         0 non-null      float64
 7   runtimeMinutes  94229 non-null  int64  
 8   genres          94229 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 7.2+ MB


In [99]:
akas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1342190 entries, 0 to 1342189
Data columns (total 8 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   titleId          1342190 non-null  object 
 1   ordering         1342190 non-null  int64  
 2   title            1342190 non-null  object 
 3   region           1342190 non-null  object 
 4   language         3676 non-null     object 
 5   types            963271 non-null   object 
 6   attributes       44717 non-null    object 
 7   isOriginalTitle  1340815 non-null  float64
dtypes: float64(1), int64(1), object(6)
memory usage: 81.9+ MB


In [98]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 475727 entries, 0 to 1262965
Data columns (total 3 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   tconst         475727 non-null  object 
 1   averageRating  475727 non-null  float64
 2   numVotes       475727 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 14.5+ MB


# Specifications - Financial Data - TMDB

In [9]:
with open('C:/Program Files/Git/.secret/tmdb_api.json', 'r') as f:
    login = json.load(f)
## Display the keys of the loaded dict # tmdb_api.json
login.keys()

dict_keys(['client-id', 'api-key'])

In [10]:
tmdb.API_KEY =  login['api-key']

In [11]:
#making function for search of movie with ID and includes movie certification
def get_movie_with_rating(movie_id):
    movie = tmdb.Movies(movie_id)
    movie_info = movie.info()
    releases = movie.releases()
    for c in releases['countries']:
        if c['iso_3166_1'] == 'US':
            movie_info['certification'] = c['certification']
    return movie_info

In [12]:
test = get_movie_with_rating("tt0848228") #testing id 1 movie
test

{'adult': False,
 'backdrop_path': '/nNmJRkg8wWnRmzQDe2FwKbPIsJV.jpg',
 'belongs_to_collection': {'id': 86311,
  'name': 'The Avengers Collection',
  'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
  'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
 'budget': 220000000,
 'genres': [{'id': 878, 'name': 'Science Fiction'},
  {'id': 28, 'name': 'Action'},
  {'id': 12, 'name': 'Adventure'}],
 'homepage': 'https://www.marvel.com/movies/the-avengers',
 'id': 24428,
 'imdb_id': 'tt0848228',
 'original_language': 'en',
 'original_title': 'The Avengers',
 'overview': 'When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director of the international peacekeeping agency known as S.H.I.E.L.D., finds himself in need of a team to pull the world back from the brink of disaster. Spanning the globe, a daring recruitment effort begins!',
 'popularity': 238.0,
 'poster_path': '/RYMX2wcKCBAr24UyPD7xwmjaTn.jpg',
 'production_companies': [{'id': 420,
   'logo_path':

In [13]:
pd.DataFrame(data=[test])

adult                     backdrop_path  \
0  False  /nNmJRkg8wWnRmzQDe2FwKbPIsJV.jpg   

                               belongs_to_collection     budget  \
0  {'id': 86311, 'name': 'The Avengers Collection...  220000000   

                                              genres  \
0  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   

                                     homepage     id    imdb_id  \
0  https://www.marvel.com/movies/the-avengers  24428  tt0848228   

  original_language original_title  ...     revenue  runtime  \
0                en   The Avengers  ...  1518815515      143   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                   tagline         title  video  vote_average vote_count  \
0  Some assembly required.  The Avengers  False         7.706      27403   

  certification  
0         PG-13  

[1 rows x 26 columns]

In [14]:
test2 = get_movie_with_rating("tt0332280") #testing id 2 movie
test2

{'adult': False,
 'backdrop_path': '/qom1SZSENdmHFNZBXbtJAU0WTlC.jpg',
 'belongs_to_collection': None,
 'budget': 29000000,
 'genres': [{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}],
 'homepage': 'http://www.newline.com/properties/notebookthe.html',
 'id': 11036,
 'imdb_id': 'tt0332280',
 'original_language': 'en',
 'original_title': 'The Notebook',
 'overview': "An epic love story centered around an older man who reads aloud to a woman with Alzheimer's. From a faded notebook, the old man's words bring to life the story about a couple who is separated by World War II, and is then passionately reunited, seven years later, after they have taken different paths.",
 'popularity': 70.325,
 'poster_path': '/rNzQyW4f8B8cQeg7Dgj3n6eT5k9.jpg',
 'production_companies': [{'id': 12,
   'logo_path': '/iaYpEp3LQmb8AfAtmTvpqd4149c.png',
   'name': 'New Line Cinema',
   'origin_country': 'US'},
  {'id': 1565, 'logo_path': None, 'name': 'Avery Pix', 'origin_country': 'US'},
  {'id': 26

In [15]:
pd.DataFrame(data=[test2])

adult                     backdrop_path belongs_to_collection    budget  \
0  False  /qom1SZSENdmHFNZBXbtJAU0WTlC.jpg                  None  29000000   

                                              genres  \
0  [{'id': 10749, 'name': 'Romance'}, {'id': 18, ...   

                                            homepage     id    imdb_id  \
0  http://www.newline.com/properties/notebookthe....  11036  tt0332280   

  original_language original_title  ...    revenue  runtime  \
0                en   The Notebook  ...  115603229      123   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                     tagline         title  video  \
0  Behind every great love is a great story.  The Notebook  False   

   vote_average vote_count certification  
0          7.88       9746         PG-13  

[1 rows x 26 columns]

# Breaking down by year 2000-2001

In [16]:
FOLDER = "Data"
os.makedirs(FOLDER, exist_ok=True)
os.listdir(FOLDER)

['Datafinal_tmdb_data_2000.csv.gz',
 'Datafinal_tmdb_data_2001.csv.gz',
 'Datatmdb_api_results_2000.json',
 'Datatmdb_api_results_2001.json',
 'title_akas.csv.gz',
 'title_basics.csv.gz',
 'title_ratings.csv.gz']

In [49]:
YEARS_TO_GET=[2000,2001]

In [50]:
def write_json(new_data, filename):   

    with open(filename,'r+') as file:
        # First we load existing data into a dict.
        file_data = json.load(file)
        ## Choose extend or append
        if (type(new_data) == list) & (type(file_data) == list):
            file_data.extend(new_data)
        else:
             file_data.append(new_data)
        # Sets file's current position at offset.
        file.seek(0)
        # convert back to json.
        json.dump(file_data, file)

In [52]:
# Start of OUTER loop for saving each year 2000-2001 (Testing)
    #"""Adapted from: https://www.geeksforgeeks.org/append-to-json-file-using-python/"""  
for YEAR in tqdm_notebook(YEARS_TO_GET,desc='YEARS',position=0):
    #Defining the JSON file to store results for year
    JSON_FILE = f'{FOLDER}tmdb_api_results_{YEAR}.json'
    # Check if file exists
    file_exists = os.path.isfile(JSON_FILE)
    # save an empty dict with just "imdb_id" to the new json file.
    # If it does not exist: create it
    if file_exists == False:
    # save an empty dict with just "imdb_id" to the new json file.
        with open(JSON_FILE,'w') as f:
            json.dump([{'imdb_id':0}],f)
                    
    # Loading in the dataframe from project part 1 as basics:

    #Saving new year as the current df
    df = basics.loc[basics['startYear']==YEAR].copy()
    # saving movie ids to list
    movie_ids = df['tconst'].copy()
    # Load existing data from json into a dataframe called "previous_df"
    previous_df = pd.read_json(JSON_FILE)
    # filtering out any ids that are already in the JSON_FILE
    movie_ids_to_get = movie_ids[~movie_ids.isin(previous_df['imdb_id'])]

# INNER Loop
    for movie_id in tqdm_notebook(movie_ids_to_get,
                                  desc=f'Movies from {YEAR}',
                                  position=1,
                                  leave=True):
                # Attempt to retrieve then data for the movie id
        try:
            temp = get_movie_with_rating(movie_id)  #This uses your pre-made function
            # Append/extend results to existing file using a pre-made function
            write_json(temp,JSON_FILE)
            # Short 20 ms sleep to prevent overwhelming server
            time.sleep(0.02)
            
                    # If it fails,  make a dict with just the id and None for certification.
        except Exception as e:
            continue
            
        final_year_df = pd.read_json(JSON_FILE)
        final_year_df.to_csv(f"{FOLDER}final_tmdb_data_{YEAR}.csv.gz",  
                             compression="gzip", index=False)
        
        final_year_df = pd.read_json(JSON_FILE)  
        final_year_df.to_csv(f"{FOLDER}final_tmdb_data_{YEAR}.csv.gz", 
                             compression="gzip", index=False)

YEARS:   0%|          | 0/2 [00:00<?, ?it/s]

Movies from 2000:   0%|          | 0/1504 [00:00<?, ?it/s]

Movies from 2001:   0%|          | 0/1605 [00:00<?, ?it/s]

# Exploratory Data Analysis

In [53]:
tmdb_2000 = pd.read_csv("Data/Datafinal_tmdb_data_2000.csv.gz", low_memory=False)
tmdb_2001 = pd.read_csv("Data/Datafinal_tmdb_data_2001.csv.gz", low_memory=False)

In [54]:
tmdb_2000.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
2  tt0113092    0.0                               NaN                   NaN   
3  tt0116391    0.0                               NaN                   NaN   
4  tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg                   NaN   

       budget                                             genres  \
0         NaN                                                NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   
4    150000.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               homepage        id original_language  \
0                                   NaN       NaN               NaN   
1                                   NaN   62127.0                en   
2                                   NaN  110977.0                en   
3                                   NaN  442869.0                hi   
4  http://www.wkw-inthemoodforlove.com/     843.0                cn   

    original_title  ...     revenue  runtime  \
0              NaN  ...         NaN      NaN   
1  The Fantasticks  ...         0.0     86.0   
2    For the Cause  ...         0.0    100.0   
3             Gang  ...         0.0    152.0   
4             花樣年華  ...  12854953.0     99.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Cantonese', 'iso_639_1': 'c...  Released   

                                             tagline                 title  \
0                                                NaN                   NaN   
1      Try to remember the first time magic happened       The Fantasticks   
2       The ultimate showdown on a forbidden planet.         For the Cause   
3                                                NaN                  Gang   
4  Feel the heat, keep the feeling burning, let t...  In the Mood for Love   

   video  vote_average vote_count certification  
0    NaN           NaN        NaN           NaN  
1    0.0           5.5       22.0           NaN  
2    0.0           5.1        8.0           NaN  
3    0.0           0.0        0.0           NaN  
4    0.0           8.1     1950.0            PG  

[5 rows x 26 columns]

In [55]:
tmdb_2001.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0035423    0.0  /ab5yL8zgRotrICzGbEl10z24N71.jpg                   NaN   
2  tt0114447    0.0                               NaN                   NaN   
3  tt0118154    0.0                               NaN                   NaN   
4  tt0118589    0.0  /9NZAirJahVilTiDNCHLFcdkwkiy.jpg                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1  48000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 14, 'nam...      NaN   
2         0.0  [{'id': 53, 'name': 'Thriller'}, {'id': 28, 'n...      NaN   
3         0.0                [{'id': 99, 'name': 'Documentary'}]      NaN   
4  22000000.0  [{'id': 18, 'name': 'Drama'}, {'id': 10402, 'n...      NaN   

         id original_language    original_title  ...     revenue  runtime  \
0       NaN               NaN               NaN  ...         NaN      NaN   
1   11232.0                en    Kate & Leopold  ...  76019048.0    118.0   
2  151007.0                en  The Silent Force  ...         0.0     90.0   
3  292980.0                en       White Hotel  ...         0.0     90.0   
4   10696.0                en           Glitter  ...   5271666.0    104.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Amharic', 'iso_639_1': 'am'...  Released   
4  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                             tagline             title  video  \
0                                                NaN               NaN    NaN   
1  If they lived in the same century they'd be pe...    Kate & Leopold    0.0   
2  They left him for dead... They should have fin...  The Silent Force    0.0   
3                                                NaN       White Hotel    0.0   
4   In music she found her dream, her love, herself.           Glitter    0.0   

   vote_average vote_count certification  
0           NaN        NaN           NaN  
1           6.3     1137.0         PG-13  
2           5.0        3.0           NaN  
3           0.0        0.0           NaN  
4           4.6      117.0         PG-13  

[5 rows x 26 columns]

In [135]:
## Merging 2 df with the same column name imdb_id
tmdb_2000_2001 = pd.concat([tmdb_2000, tmdb_2001], ignore_index=True)

In [136]:
tmdb_2000_2001.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0113026    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
2  tt0113092    0.0                               NaN                   NaN   
3  tt0116391    0.0                               NaN                   NaN   
4  tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg                   NaN   

       budget                                             genres  \
0         NaN                                                NaN   
1  10000000.0  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
2         0.0           [{'id': 878, 'name': 'Science Fiction'}]   
3         0.0  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   
4    150000.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   

                               homepage        id original_language  \
0                                   NaN       NaN               NaN   
1                                   NaN   62127.0                en   
2                                   NaN  110977.0                en   
3                                   NaN  442869.0                hi   
4  http://www.wkw-inthemoodforlove.com/     843.0                cn   

    original_title  ...     revenue  runtime  \
0              NaN  ...         NaN      NaN   
1  The Fantasticks  ...         0.0     86.0   
2    For the Cause  ...         0.0    100.0   
3             Gang  ...         0.0    152.0   
4             花樣年華  ...  12854953.0     99.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Cantonese', 'iso_639_1': 'c...  Released   

                                             tagline                 title  \
0                                                NaN                   NaN   
1      Try to remember the first time magic happened       The Fantasticks   
2       The ultimate showdown on a forbidden planet.         For the Cause   
3                                                NaN                  Gang   
4  Feel the heat, keep the feeling burning, let t...  In the Mood for Love   

   video  vote_average vote_count certification  
0    NaN           NaN        NaN           NaN  
1    0.0           5.5       22.0           NaN  
2    0.0           5.1        8.0           NaN  
3    0.0           0.0        0.0           NaN  
4    0.0           8.1     1950.0            PG  

[5 rows x 26 columns]

# How many movies had at least some valid financial information (values > 0 for budget OR revenue)?


In [141]:
tmdb_2000_2001[(tmdb_2000_2001['budget'] > 0.0) | (tmdb_2000_2001['revenue'] > 0.0)].count()

imdb_id                  629
adult                    629
backdrop_path            521
belongs_to_collection    102
budget                   629
genres                   629
homepage                  82
id                       629
original_language        629
original_title           629
overview                 626
popularity               629
poster_path              605
production_companies     629
production_countries     629
release_date             629
revenue                  629
runtime                  629
spoken_languages         629
status                   629
tagline                  482
title                    629
video                    629
vote_average             629
vote_count               629
certification            425
dtype: int64

- 629 movies had budget or revenue greater than 0.

- Please exclude any movies with 0's for budget AND revenue from the remaining visualizations.

In [147]:
tmdb_2000_2001 = tmdb_2000_2001[(tmdb_2000_2001['budget'] > 0.0) & (tmdb_2000_2001['revenue'] > 0.0)]
tmdb_2000_2001 

imdb_id  adult                     backdrop_path  \
4     tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg   
11    tt0120467    0.0  /knok3mNReKqPTplnnqz7E4dd7mD.jpg   
13    tt0120630    0.0  /sPAwM8WxMdXNlqeP4F9DMe9LYoY.jpg   
16    tt0120753    0.0                               NaN   
17    tt0120755    0.0  /24DZfupDlhXeTchmcOkoGRhP5Vg.jpg   
...         ...    ...                               ...   
2297  tt0294289    0.0                               NaN   
2329  tt0296042    0.0  /aw0EeseSr10dVUuY6Aiv6HhP8UW.jpg   
2436  tt0310790    0.0  /hAvu0oWrEbbPN71QC8w1NIoUYQT.jpg   
2549  tt0445841    0.0                               NaN   
2568  tt1039952    0.0  /yTGmGjAd0gfpVK0ezoqWgx7AJkp.jpg   

                                  belongs_to_collection       budget  \
4                                                   NaN     150000.0   
11                                                  NaN     120000.0   
13    {'id': 718551, 'name': 'Chicken Run Collection...   45000000.0   
16                                                  NaN    8000000.0   
17    {'id': 87359, 'name': 'Mission: Impossible Col...  125000000.0   
...                                                 ...          ...   
2297                                                NaN        344.0   
2329                                                NaN    1400010.0   
2436  {'id': 518467, 'name': 'El Bosque Animado', 'p...    3747230.0   
2549                                                NaN   12500000.0   
2568                                                NaN     500000.0   

                                                 genres  \
4     [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
11    [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   
13    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
16    [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   
17    [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
...                                                 ...   
2297  [{'id': 35, 'name': 'Comedy'}, {'id': 10749, '...   
2329  [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   
2436                  [{'id': 16, 'name': 'Animation'}]   
2549  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
2568  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   

                                  homepage        id original_language  \
4     http://www.wkw-inthemoodforlove.com/     843.0                cn   
11                                     NaN   19085.0                en   
13                                     NaN    7443.0                en   
16                                     NaN     318.0                en   
17                                     NaN     955.0                en   
...                                    ...       ...               ...   
2297          http://www.americanadobo.com   50819.0                en   
2329                                   NaN    9696.0                ja   
2436                                   NaN   49612.0                gl   
2549                                   NaN  233308.0                zh   
2568                                   NaN   69399.0                ta   

                original_title  ...      revenue  runtime  \
4                         花樣年華  ...   12854953.0     99.0   
11                      Vulgar  ...      14904.0     87.0   
13                 Chicken Run  ...  224834564.0     84.0   
16    The Million Dollar Hotel  ...     105983.0    122.0   
17      Mission: Impossible II  ...  546388105.0    123.0   
...                        ...  ...          ...      ...   
2297            American Adobo  ...          4.0    104.0   
2329                      殺し屋1  ...      80631.0    129.0   
2436          O Bosque Animado  ...    1951816.0     83.0   
2549                   一个烂赌的传说  ...  123021750.0     94.0   
2568                 சிட்டிசன்  ...     500000.0    157.0   

                                       spoken_languages    status  \
4     [{'english_name'

# How many movies are there in each of the certification categories (G/PG/PG-13/R)?

In [148]:
#movies in each of the certifications
tmdb_2000_2001['certification'].value_counts()

R        140
PG-13    117
PG        29
G         13
NR         6
Name: certification, dtype: int64

# What is the average revenue per certification category?

In [151]:
#average revenue per certification 
tmdb_2000_2001.groupby('certification')['revenue'].mean()

certification
G        1.332169e+08
NR       2.382332e+07
PG       1.346138e+08
PG-13    1.107121e+08
R        5.344654e+07
Name: revenue, dtype: float64

# What is the average budget per certification category?

In [152]:
#average budget per certification 
tmdb_2000_2001.groupby('certification')['budget'].mean()

certification
G        4.400000e+07
NR       1.478333e+07
PG       5.374138e+07
PG-13    4.680682e+07
R        2.750255e+07
Name: budget, dtype: float64

# Deliverables

In [153]:
#saving df to csv file 
tmdb_2000_2001.to_csv("Data/tmdb_results_combined.csv.gz")

# Normalizing & Creating MySQL database

In [6]:
one.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351 entries, 0 to 350
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             351 non-null    int64  
 1   imdb_id                351 non-null    object 
 2   adult                  351 non-null    float64
 3   backdrop_path          338 non-null    object 
 4   belongs_to_collection  84 non-null     object 
 5   budget                 351 non-null    float64
 6   genres                 351 non-null    object 
 7   homepage               57 non-null     object 
 8   id                     351 non-null    float64
 9   original_language      351 non-null    object 
 10  original_title         351 non-null    object 
 11  overview               351 non-null    object 
 12  popularity             351 non-null    float64
 13  poster_path            350 non-null    object 
 14  production_companies   351 non-null    object 
 15  produc

In [125]:
one = pd.read_csv('Data/tmdb_results_combined.csv.gz')
one.head(2)

Unnamed: 0    imdb_id  adult                     backdrop_path  \
0           4  tt0118694    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg   
1          11  tt0120467    0.0  /knok3mNReKqPTplnnqz7E4dd7mD.jpg   

  belongs_to_collection    budget  \
0                   NaN  150000.0   
1                   NaN  120000.0   

                                              genres  \
0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
1  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   

                               homepage       id original_language  ...  \
0  http://www.wkw-inthemoodforlove.com/    843.0                cn  ...   
1                                   NaN  19085.0                en  ...   

      revenue runtime                                   spoken_languages  \
0  12854953.0    99.0  [{'english_name': 'Cantonese', 'iso_639_1': 'c...   
1     14904.0    87.0  [{'english_name': 'English', 'iso_639_1': 'en'...   

     status                                            tagline  \
0  Released  Feel the heat, keep the feeling burning, let t...   
1  Released   Everyone loves a clown... some more than others.   

                  title video  vote_average  vote_count certification  
0  In the Mood for Love   0.0           8.1      1950.0            PG  
1                Vulgar   0.0           5.4        42.0             R  

[2 rows x 27 columns]

In [7]:
basicsOne = pd.read_csv("Data/title_basics.csv.gz", low_memory=False)

In [9]:
basicsOne.head(2)

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   

                               originalTitle  isAdult  startYear  endYear  \
0                             Kate & Leopold        0       2001      NaN   
1  El Tango del Viudo y Su Espejo Deformante        0       2020      NaN   

   runtimeMinutes                  genres  
0             118  Comedy,Fantasy,Romance  
1              70                   Drama

In [8]:
ratingOne = pd.read_csv("Data/title_ratings.csv.gz", low_memory=False)

In [10]:
ratingOne.head(2)

tconst  averageRating  numVotes
0  tt0000001            5.7      1905
1  tt0000002            5.9       256

In [12]:
#splitting genre from list to column
basicsOne['genres_split'] = basicsOne['genres'].str.split(',')

In [20]:
#separate the list of genres into new rows
exploded_genres = basicsOne.explode('genres_split')
exploded_genres

tconst titleType                                       primaryTitle  \
0      tt0035423     movie                                     Kate & Leopold   
0      tt0035423     movie                                     Kate & Leopold   
0      tt0035423     movie                                     Kate & Leopold   
1      tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2      tt0069049     movie                         The Other Side of the Wind   
...          ...       ...                                                ...   
94227  tt9916190     movie                                          Safeguard   
94227  tt9916190     movie                                          Safeguard   
94227  tt9916190     movie                                          Safeguard   
94228  tt9916362     movie                                              Coven   
94228  tt9916362     movie                                              Coven   

                                   originalTitle  isAdult  startYear  endYear  \
0                                 Kate & Leopold        0       2001      NaN   
0                                 Kate & Leopold        0       2001      NaN   
0                                 Kate & Leopold        0       2001      NaN   
1      El Tango del Viudo y Su Espejo Deformante        0       2020      NaN   
2                     The Other Side of the Wind        0       2018      NaN   
...                                          ...      ...        ...      ...   
94227                                  Safeguard        0       2020      NaN   
94227                                  Safeguard        0       2020      NaN   
94227                                  Safeguard        0       2020      NaN   
94228                                   Akelarre        0       2020      NaN   
94228                                   Akelarre        0       2020      NaN   

       runtimeMinutes                     genres genres_split  
0                 118     Comedy,Fantasy,Romance       Comedy  
0                 118     Comedy,Fantasy,Romance      Fantasy  
0                 118     Comedy,Fantasy,Romance      Romance  
1                  70                      Drama        Drama  
2                 122                      Drama        Drama  
...               ...                        ...          ...  
94227              95  Action,Adventure,Thriller       Action  
94227              95  Action,Adventure,Thriller    Adventure  
94227              95  Action,Adventure,Thriller     Thriller  
94228              92              Drama,History        Drama  
94228              92              Drama,History      History  

[183572 rows x 10 columns]

In [21]:
#using explode to only take unique entries
genres_split = basicsOne['genres'].str.split(',')
unique_genres = genres_split.explode().unique()
unique_genres

array(['Comedy', 'Fantasy', 'Romance', 'Drama', 'Horror', 'Sci-Fi',
       'Mystery', 'Musical', 'Action', 'Adventure', 'Crime', 'Thriller',
       'Biography', 'Documentary', 'Music', 'Animation', 'Family',
       'History', 'War', 'Sport', 'Western', 'Adult', 'Short',
       'Reality-TV', 'News', 'Talk-Show', 'Game-Show'], dtype=object)

In [22]:
unique_genres = sorted(exploded_genres['genres_split'].unique())

In [25]:
#creating new table title_genres with id and separated genres
title_genres = exploded_genres[['tconst', 'genres_split']].copy()
title_genres.head()

tconst genres_split
0  tt0035423       Comedy
0  tt0035423      Fantasy
0  tt0035423      Romance
1  tt0062336        Drama
2  tt0069049        Drama

In [26]:
#Creating a genre mapper dictionary to replace string genres with integers
genre_ints = range(len(unique_genres))
genre_id_map = dict(zip(unique_genres, genre_ints))
genre_id_map

{'Action': 0,
 'Adult': 1,
 'Adventure': 2,
 'Animation': 3,
 'Biography': 4,
 'Comedy': 5,
 'Crime': 6,
 'Documentary': 7,
 'Drama': 8,
 'Family': 9,
 'Fantasy': 10,
 'Game-Show': 11,
 'History': 12,
 'Horror': 13,
 'Music': 14,
 'Musical': 15,
 'Mystery': 16,
 'News': 17,
 'Reality-TV': 18,
 'Romance': 19,
 'Sci-Fi': 20,
 'Short': 21,
 'Sport': 22,
 'Talk-Show': 23,
 'Thriller': 24,
 'War': 25,
 'Western': 26}

In [27]:
#replacing string_genre to genre_id using genre_id_map
title_genres['genre_id'] = title_genres['genres_split'].map(genre_id_map)

In [126]:
#TABLE 1
#dropping genres_split and keeping genre_id 
title_genres = title_genres.drop(columns='genres_split')
title_genres.head()

KeyError: "['genres_split'] not found in axis"

In [134]:
#making table 1 into sql data with no index
title_genres.to_sql('title_genres', engine, if_exists='replace', index=False)

183572

In [135]:
#checking table 1
q = ''' SELECT * FROM title_genres LIMIT 5;

 '''

pd.read_sql(q, engine)

tconst  genre_id
0  tt0035423         5
1  tt0035423        10
2  tt0035423        19
3  tt0062336         8
4  tt0069049         8

In [136]:
#TABLE 2
#converting genre_map_id into table
genres = pd.DataFrame({'genre_id': genre_id_map.values(),
               'genre_name': genre_id_map.keys()})
genres.head(2)

genre_id genre_name
0         0     Action
1         1      Adult

In [137]:
#Table 2 schema
# Calculating max string lengths for object columns
name_len = genres['genre_name'].fillna('').map(len).max()
## Creating a schema dictonary using Sqlalchemy datatype objects
genres_schema = {
    'genre_id': Integer(), 
    "genre_name": Text(name_len+1)}

In [138]:
#Running df.to_sql with the dtype argument
genres.to_sql('genres', engine, dtype=genres_schema, 
                 if_exists='replace', index=False)

27

In [139]:
#adding primary key to Table 2
engine.execute('ALTER TABLE genres ADD PRIMARY KEY (genre_id)')

In [140]:
#checking table 2
q = ''' SELECT * FROM genres LIMIT 5;

 '''

pd.read_sql(q, engine)

genre_id genre_name
0         0     Action
1         1      Adult
2         2  Adventure
3         3  Animation
4         4  Biography

In [141]:
#Table 3
title_basics = basicsOne[['tconst', 'primaryTitle', 'startYear',
                                   'endYear', 'runtimeMinutes']]
title_basics.head(2)

tconst                                       primaryTitle  startYear  \
0  tt0035423                                     Kate & Leopold       2001   
1  tt0062336  The Tango of the Widower and Its Distorting Mi...       2020   

   endYear  runtimeMinutes  
0      NaN             118  
1      NaN              70

In [142]:
# getting max string length for tconst in title_genres
max_str_len = title_genres['tconst'].fillna('').map(len).max()
max_str_len

10

In [143]:
#Table 3 schema
# Calculating max string lengths for object columns
tkey_len = basicsOne['tconst'].fillna('').map(len).max()
title_len = basicsOne['primaryTitle'].fillna('').map(len).max()
## Creating a schema dictonary using Sqlalchemy datatype objects
title_basics_schema = {
    "tconst": String(tkey_len+1), 
    "primaryTitle": Text(title_len+1),
    'startYear':Float(),
    'endYear':Float(),
    'runtimeMinutes':Integer()}

In [144]:
#Running df.to_sql with the dtype argument
title_basics.to_sql('title_basics', engine, dtype=title_basics_schema, 
                 if_exists='replace', index=False)

94229

In [145]:
#adding primary key to Table 3
engine.execute('ALTER TABLE title_basics ADD PRIMARY KEY (tconst)')

In [146]:
#checking table 3
q = ''' SELECT * FROM title_basics LIMIT 5;

 '''

pd.read_sql(q, engine)

tconst                                       primaryTitle  startYear  \
0  tt0035423                                     Kate & Leopold     2001.0   
1  tt0062336  The Tango of the Widower and Its Distorting Mi...     2020.0   
2  tt0069049                         The Other Side of the Wind     2018.0   
3  tt0088751                                  The Naked Monster     2005.0   
4  tt0094859                                         Chief Zabu     2016.0   

  endYear  runtimeMinutes  
0    None             118  
1    None              70  
2    None             122  
3    None             100  
4    None              74

In [147]:
#Table 4
title_ratings = ratingOne[['tconst', 'averageRating', 'numVotes']]
title_ratings.head(2)

tconst  averageRating  numVotes
0  tt0000001            5.7      1905
1  tt0000002            5.9       256

In [148]:
#Table 4 schema
# Calculating max string lengths for object columns
rkey_len = ratingOne['tconst'].fillna('').map(len).max()
## Creating a schema dictonary using Sqlalchemy datatype objects
ratings_schema = {
    "tconst": String(rkey_len+1),
    'averageRating':Float(),
    'numVotes':Integer()}

In [149]:
#Running df.to_sql with the dtype argument
title_ratings.to_sql('title_ratings', engine, dtype=ratings_schema, 
                 if_exists='replace', index=False)

475727

In [150]:
#adding primary key to Table 4
engine.execute('ALTER TABLE title_ratings ADD PRIMARY KEY (tconst)')

In [151]:
#checking table 4
q = ''' SELECT * FROM title_ratings LIMIT 5;

 '''

pd.read_sql(q, engine)

tconst  averageRating  numVotes
0  tt0000001            5.7      1905
1  tt0000002            5.9       256
2  tt0000005            6.2      2517
3  tt0000006            5.2       173
4  tt0000007            5.4       783

In [152]:
#Table 5
tmdb_data = one[['imdb_id', 'revenue', 'budget', 'certification']]
tmdb_data.head(2)

imdb_id     revenue    budget certification
0  tt0118694  12854953.0  150000.0            PG
1  tt0120467     14904.0  120000.0             R

In [153]:
#Table 5 schema
# Calculating max string lengths for object columns
tmkey_len = tmdb_data['imdb_id'].fillna('').map(len).max()
tmname_len = tmdb_data['certification'].fillna('').map(len).max()
## Creating a schema dictonary using Sqlalchemy datatype objects
tmdb_data_schema = {
    "tconst": String(tmkey_len+1),
    'revenue':Float(),
    'budget':Float(),
    "certification": Text(tmname_len+1)}

In [154]:
#Running df.to_sql with the dtype argument for table 5
tmdb_data.to_sql('tmdb_data', engine, dtype=tmdb_data_schema, 
                 if_exists='replace', index=False)

351

In [155]:
#checking table 5
q = ''' SELECT * FROM tmdb_data LIMIT 5;

 '''

pd.read_sql(q, engine)

imdb_id      revenue       budget certification
0  tt0118694   12855000.0     150000.0            PG
1  tt0120467      14904.0     120000.0             R
2  tt0120630  224835000.0   45000000.0             G
3  tt0120753     105983.0    8000000.0             R
4  tt0120755  546388000.0  125000000.0         PG-13

In [156]:
q = ''' SHOW TABLES;

 '''

pd.read_sql(q, engine)

Tables_in_movies
0           genres
1     title_basics
2     title_genres
3    title_ratings
4        tmdb_data